src/esp32/hcsr04.rs

```diff
@@ -0,0 +1,289 @@
+// Support for HC-SR04 style ultrasonic ranging modules. See
+// https://cdn.sparkfun.com/datasheets/Sensors/Proximity/HCSR04.pdf.
+
+use std::{
+    collections::HashMap,
+    num::NonZeroU32,
+    sync::{
+        atomic::{AtomicI64, Ordering},
+        Arc, Mutex,
+    },
+    time::Duration,
+};
+
+use crate::{
+    common::{
+        board::BoardType,
+        config::{AttributeError, ConfigType},
+        registry::{self, ComponentRegistry, Dependency},
+        sensor::{
+            GenericReadingsResult, Readings, Sensor, SensorResult, SensorT, SensorType,
+            TypedReadingsResult,
+        },
+        status::Status,
+    },
+    google, DoCommand,
+};
+
+use esp_idf_hal::{
+    delay::TickType,
+    gpio::{
+        enable_isr_service, init_isr_alloc_flags, AnyIOPin, Input, InterruptType, Output,
+        PinDriver, Pull,
+    },
+    task::notification::{Notification, Notifier},
+};
+
+use esp_idf_sys::{esp, gpio_isr_handler_add, gpio_isr_handler_remove};
+
+pub(crate) fn register_models(registry: &mut ComponentRegistry) {
+    if registry
+        .register_sensor("ultrasonic", &HCSR04Sensor::from_config)
+        .is_err()
+    {
+        log::error!("HCSR04Sensor is already registered");
+    }
+}
+
+struct IsrSharedState {
+    // The state machine used to track interrupts and compute the
+    // length of the echo pulse. It holds one of the following values:
+    //
+    // 0: Starting state, ready to take a reading
+    // i64 > 0: Millisecond timestamp of first edge of echo signal
+    // -1 : Lockout state, further interrupts will be ignored
```

Comment:
I'm not 100% convinced this state is needed, and managing it costs an extra compare_and_exchange in the isr. I added it while debugging state handling around timeouts, but there were other, subsequently resolved, issues that may have been the actual cause. I will re-test with it removed.

Reply:
The lockout has been removed, further testing indicated it was not required.